File: src/Candidato.java
import java.time.LocalDate;
import java.time.Period;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;

public class Candidato {
    private int votos;
    private int numero;
    private String nomeUrna;
    private Partido partido;
    private LocalDate dataNascimento;
    private Genero genero;
    private SituacaoEleitoral situacaoEleitoral;

    /**
     * Construtor de candidato.
     * @param numero Número do candidato na eleição.
     * @param nome Nome do candidato na urna.
     * @param partido Partido do candidato.
     * @param dataNasc Data de nascimento do candidato (formato: dd/MM/yyyy).
     * @param genero Código do gênero do candidato.
     * @param situacao Código da situação eleitoral do candidato.
     * @throws IllegalArgumentException Se o partido for nulo ou se a data de nascimento for inválida.
     */
    public Candidato(int numero, String nome, Partido partido, String dataNasc, int genero, int situacao) {
        if (partido == null) throw new IllegalArgumentException("Partido não pode ser nulo");

        this.numero = numero;
        this.nomeUrna = nome;
        this.partido = partido;

        // Valida a data de nascimento
        try {
            this.dataNascimento = LocalDate.parse(dataNasc, DateTimeFormatter.ofPattern("dd/MM/yyyy"));
        } catch (DateTimeParseException e) {
            throw new IllegalArgumentException("Data de nascimento inválida: " + dataNasc);
        }

        this.genero = Genero.fromCodigo(genero);
        this.situacaoEleitoral = SituacaoEleitoral.fromCodigo(situacao);
    }

    /**
     * Retorna o número do candidato.
     * @return Número do candidato.
     */
    public int getNumero() {
        return this.numero;
    }

    /**
     * Retorna o nome do candidato na urna. Se o partido for federado, adiciona um asterisco ao nome.
     * @param flagFederado Indica se o partido é federado ou não.
     * @return Nome do candidato.
     */
    public String getNomeUrna(int flagFederado) {
        // Se o partido for federado e flagFederado for 1, retorna nome com asterisco
        if (this.partido.getNumeroFederacao() != -1 && flagFederado != 0)
            return "*" + this.nomeUrna;
        return nomeUrna;
    }

    /**
     * Retorna o partido do candidato.
     * @return Partido do candidato.
     */
    public Partido getPartido() {
        return this.partido;
    }

    /**
     * Retorna o número de votos do candidato.
     * @return Número de votos do candidato.
     */
    public int getVotos() {
        return this.votos;
    }

    /**
     * Adiciona votos ao candidato. Não permite votos negativos.
     * @param votos Quantidade de votos a ser adicionada.
     * @throws IllegalArgumentException Se a quantidade de votos for negativa.
     */
    public void addVotos(int votos) {
        if (votos < 0) throw new IllegalArgumentException("Votos não podem ser negativos");
        this.votos += votos;
    }

    /**
     * Retorna o gênero do candidato.
     * @return Gênero do candidato.
     */
    public Genero getGenero() {
        return this.genero;
    }

    /**
     * Retorna a data de nascimento do candidato.
     * @return Data de nascimento do candidato.
     */
    public LocalDate getDataNascimento() {
        return this.dataNascimento;
    }

    /**
     * Verifica se o candidato foi eleito.
     * @return Verdadeiro se o candidato for eleito, caso contrário falso.
     */
    public boolean isEleito() {
        return this.situacaoEleitoral == SituacaoEleitoral.ELEITO ||
               this.situacaoEleitoral == SituacaoEleitoral.ELEITO_POR_QUOCIENTE_PARTIDARIO ||
               this.situacaoEleitoral == SituacaoEleitoral.ELEITO_POR_MEDIA;
    }

    /**
     * Calcula a idade do candidato na data da eleição.
     * @param dataEleicaoStr Data da eleição (formato: dd/MM/yyyy).
     * @return Idade do candidato na data da eleição.
     * @throws IllegalArgumentException Se a data da eleição for inválida.
     */
    public int getIdade(String dataEleicaoStr) {
        LocalDate dataEleicao;
        try {
            dataEleicao = LocalDate.parse(dataEleicaoStr, DateTimeFormatter.ofPattern("dd/MM/yyyy"));
        } catch (DateTimeParseException e) {
            throw new IllegalArgumentException("Data de eleição inválida: " + dataEleicaoStr);
        }

        return Period.between(this.dataNascimento, dataEleicao).getYears();
    }
}

File: trad/Candidato..cpp
#include <sstream>
#include <iomanip>

#include "Candidato.hpp"
#include "Partido.hpp"

using namespace std;

Candidato::Candidato(const int& numero, const string& nomeUrna, Partido &partido,
                     const string &dataNascStr, const int& genero, const int& situacao)
    : numero(numero),
      nomeUrna(nomeUrna),
      partido(&partido),
      genero(Genero::fromCodigo(genero)), 
      situacaoEleitoral(SituacaoEleitoral::fromCodigo(situacao)),
      dataNascimento(DataUtils::fromString(dataNascStr))
{

    if (!(&partido))
    {
        throw invalid_argument("Partido não pode ser nulo");
    }

    this->votos = 0;
}

// Getters
int Candidato::getNumero() const { return numero; }

string Candidato::getNomeUrna(int flagFederado) const
{
    if (partido->getNumeroFederacao() != -1 && flagFederado != 0)
    {
        return "*" + nomeUrna;
    }
    return nomeUrna;
}

Partido &Candidato::getPartido() const { return *partido; }
int Candidato::getVotos() const { return votos; }
Genero Candidato::getGenero() const { return genero; }
DataUtils Candidato::getDataNascimento() const { return dataNascimento; }

bool Candidato::isEleito() const
{
    return (situacaoEleitoral.getCodigo() == SituacaoEleitoral::ELEITO ||
           situacaoEleitoral.getCodigo() == SituacaoEleitoral::ELEITO_POR_QUOCIENTE_PARTIDARIO ||
           situacaoEleitoral.getCodigo() == SituacaoEleitoral::ELEITO_POR_MEDIA);
}

void Candidato::addVotos(int votos)
{
    if (votos < 0)
    {
        throw invalid_argument("Votos não podem ser negativos");
    }
    this->votos += votos;
}

int Candidato::getIdade(const string &dataEleicaoStr) const
{
    DataUtils dataleicao = DataUtils::fromString(dataEleicaoStr);
    int result = this->dataNascimento.calculaPeriodoEmAnos(dataleicao);
    if(result < 0){
        throw invalid_argument( "Idade obtida foi negativa. Data de eleicao ou de nascimento do candidato estao incorretas.");
    }
    return result;
}